cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueue.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.schedule.queue;
+
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.proto.id.ScheduleId;
+
+import java.util.List;
+
+/**
+ * Responsible for keeping track of {@link Job}s, which correspond to schedules that have been triggered,
+ * but not yet executed.
+ */
+interface JobQueue {
+
+  /**
+   * Returns a list of Jobs associated with the given schedule Id.
+   */
+  List<Job> getJobsForSchedule(ScheduleId scheduleId);
+
+  /**
+   * Creates a new Job in the queue or updates an existing Job.
```

Comment:
Yes, that is more natural (e.g. Job.setState will make sure the change in state is preserved when transaction committed). The only thing is you have to make sure the mutable part of the job object can only be used in transaction.

Alternatively, you can structure `Job` as immutable object (I like immutable object more) and then have method on the `JobQueue` to perform mutation, which I think it enforces a better contract